subsys/net/ip/tcp.c

```diff
@@ -2157,12 +2157,18 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 			if (th_seq(th) == conn->ack) {
 				verdict = tcp_data_received(conn, pkt, &len);
 			} else if (net_tcp_seq_greater(conn->ack, th_seq(th))) {
-				tcp_out(conn, ACK); /* peer has resent */
+				if ((len > 0) || FL(&fl, &, SYN)) {
```

Comment:
Hmmm... I was not aware of this usecase, let's re-phase: how to differentitate between an ACK and a zero length packet, like a keep alive?
I looked at https://www.wireshark.org/docs/wsug_html_chunked/ChAdvTCPAnalysis.html#:~:text=TCP%20Keep%2DAlive%20ACK&text=The%20current%20sequence%20number%20is%20the%20same%20as%20the%20next,SYN%2C%20FIN%2C%20or%20RST, but did not get a very clear awnwer.

Reply:
Hmm, I might not be seeing something, but is there a scenario when a proper empty ACK is sent with an older `SEQ` number? My understanding is that an older `SEQ` would be sent in only 2 cases:
* ACK from our side was lost, so the peer retransmits,
* or peer sends this keep-alive message, with an older `SEQ` number.

In either case, such a packet should be acknowledged. https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/overview.html exaplains that the keep-alive message specifically uses an older `SEQ` to assure ACK reply in any case, even if peer doesn't support keep-alive mechanism.

But perhaps I miss some corner case, when out-of-order data processing kicks in?

Reply:
It is a though one. The risk is that the both ends of a connection end up in a situation where they keep sending ACKs on each others ACKs en you end up in an indefinite loop.

The only real reason I can think of when you will see out of order ACKs is when you will have bidirectional data flow at the same time. Then in the sending part a packet is being lost, the other side will be transmitting out of order ACKs.
Acking all the old ACKs does not pose too big of a risk for that reason.

There are only 1 special case where you want to ACK. Is a keep alive packet. The other closely related one in the ZWP. With ZWP I see most specifications state that you have to transmit a "dummy" byte. Although the current Zephyr implementation does modify the sequence number, but sends no data. It looks more like a keep alive packet for that reason.
The ACK behavior in this change will therefore not work with the ZWP for Zephyr.

The keep alive packets can only be arriving when the send_data_total = 0. Otherwise the condition will resolve itself anyway.
For the time being I'll revert this part of the change. But maybe it is at least useful to add a comment and create an issue for the ZWP behavior.

Reply:
Regarding the Keep-alive packet, I don't think that sending an empty packet is incorrect, actually [RFC 1122](https://datatracker.ietf.org/doc/html/rfc1122#page-102) states that sending dummy data byte is rather an optional behaviour:
```
An implementation SHOULD send a keep-alive segment with no
data; however, it MAY be configurable to send a keep-alive
segment containing one garbage octet, for compatibility with
erroneous TCP implementations.
```

Reply:
Indeed keep-alive packets should have a zero length, ZWP however should have a dummy byte.

Reply:
Ok, that's the theory, we can fix the Zephyr implementation of ZWP sending, although it complicates it a bit (for example what to do if there is no actual data to send). Nonetheless, in the actual field I've observed that some servers send Keep-alive message as a ZWP (that's the reason it's implemented like this in Zephyr as well) so we need to be prepared to handle this case as well.